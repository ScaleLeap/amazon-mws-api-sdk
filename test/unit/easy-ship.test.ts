import {
  createMockHttpClient,
  mockMwsServiceStatus,
  mockParsingError,
  parsingErrorRegex,
} from '../utils'

function mockEnum() {
  /**
   * Mock everything in purify-ts, restore it back to normal,
   *  except for `enumeration` which will be stubbed to accept
   * https://github.com/facebook/jest/issues/936#issuecomment-265074320
   */
  const original = jest.requireActual('purify-ts')

  return {
    ...original, // Pass down all the exported objects
    enumeration: <T extends Record<string, string | number>>(enumeration: T) => {
      const enumValues = Object.values(enumeration)

      return original.Codec.custom({
        decode: (input: string | number) => {
          if (typeof input !== 'string' && typeof input !== 'number') {
            return original.Left(`Expected enum, received ${input}`)
          }

          const enumIndex = enumValues.indexOf(input)

          return enumIndex !== -1 || input === 'String'
            ? original.Right((enumValues[enumIndex] as T[keyof T]) || 'String')
            : original.Left(`Expected enum, received ${input}`)
        },
        encode: original.identity,
        schema: () => ({ enum: enumValues }),
      })
    },
  }
}
jest.mock('purify-ts', () => mockEnum())

const mockDimensions = {
  Length: 1,
  Width: 1,
  Height: 1,
  Unit: 'cm',
}

const mockWeight = {
  Value: 1,
  Unit: 'g',
}

const mockPickupSlot = {
  SlotId: '',
  PickupTimeStart: new Date(),
  PickupTimeEnd: new Date(),
}

const mockScheduledPackageId = {
  AmazonOrderId: '',
}

describe('easyShip', () => {
  describe('getScheduledPackage', () => {
    const parameters = {
      ScheduledPackageId: mockScheduledPackageId,
      MarketplaceId: '',
    }

    it('returns info about the scheduled package if succesful', async () => {
      expect.assertions(1)

      const mockGetScheduledPackage = createMockHttpClient('easy_ship_get_scheduled_package')

      expect(
        await mockGetScheduledPackage.easyShip.getScheduledPackage(parameters),
      ).toMatchSnapshot()
    })

    it("throws a parsing error when the status response isn't valid", async () => {
      expect.assertions(1)

      await expect(() => mockParsingError.easyShip.getScheduledPackage(parameters)).rejects.toThrow(
        parsingErrorRegex,
      )
    })
  })

  describe('updateScheduledPackages', () => {
    const mockScheduledPackageUpdateDetails = {
      ScheduledPackageId: mockScheduledPackageId,
      PackagePickupSlot: mockPickupSlot,
    }
    const parameters = {
      MarketplaceId: '',
      ScheduledPackageUpdateDetailsList: [mockScheduledPackageUpdateDetails],
    }

    it('returns a list of packages if succesful', async () => {
      expect.assertions(1)

      const mockUpdateScheduledPackages = createMockHttpClient(
        'easy_ship_update_scheduled_packages',
      )

      expect(
        await mockUpdateScheduledPackages.easyShip.updateScheduledPackages(parameters),
      ).toMatchSnapshot()
    })

    it('throws a parsing error when the status response is not valid', async () => {
      expect.assertions(1)

      await expect(() =>
        mockParsingError.easyShip.updateScheduledPackages(parameters),
      ).rejects.toThrow(parsingErrorRegex)
    })
  })

  describe('createScheduledPackage', () => {
    const mockPackageRequestDetails = {
      PackagePickupSlot: mockPickupSlot,
    }
    const parameters = {
      AmazonOrderId: '',
      MarketplaceId: '',
      PackageRequestDetails: mockPackageRequestDetails,
    }

    it('returns information about the scheduled package if structure is correct', async () => {
      expect.assertions(1)

      const mockCreateSchedulePackage = createMockHttpClient(
        'easy_ship_create_scheduled_package_from_c_sharp',
      )

      expect(
        await mockCreateSchedulePackage.easyShip.createScheduledPackage(parameters),
      ).toMatchSnapshot()
    })

    it('returns information about the scheduled package if succesful', async () => {
      expect.assertions(1)

      const mockCreateSchedulePackage = createMockHttpClient('easy_ship_create_scheduled_package')

      expect(
        await mockCreateSchedulePackage.easyShip.createScheduledPackage(parameters),
      ).toMatchSnapshot()
    })

    it('throws a parsing error when the status response is not valid', async () => {
      expect.assertions(1)

      await expect(() =>
        mockParsingError.easyShip.createScheduledPackage(parameters),
      ).rejects.toThrow(parsingErrorRegex)
    })
  })

  describe('listPickupSlots', () => {
    const parameters = {
      MarketplaceId: '',
      AmazonOrderId: '',
      PackageDimensions: mockDimensions,
      PackageWeight: mockWeight,
    }

    it('returns list of pickup slots if succesful', async () => {
      expect.assertions(1)

      const mockListPickupSlots = createMockHttpClient('easy_ship_list_pickup_slots')

      expect(await mockListPickupSlots.easyShip.listPickupSlots(parameters)).toMatchSnapshot()
    })

    it('throws a parsing error when the status response isn t valid', async () => {
      expect.assertions(1)

      await expect(() => mockParsingError.easyShip.listPickupSlots(parameters)).rejects.toThrow(
        parsingErrorRegex,
      )
    })
  })

  describe('getServiceStatus', () => {
    it('returns a parsed model when the status response is valid', async () => {
      expect.assertions(1)

      expect(await mockMwsServiceStatus.easyShip.getServiceStatus()).toMatchSnapshot()
    })

    it('throws a parsing error when the status response isn t valid', async () => {
      expect.assertions(1)

      await expect(() => mockParsingError.easyShip.getServiceStatus()).rejects.toThrow(
        parsingErrorRegex,
      )
    })
  })
})
