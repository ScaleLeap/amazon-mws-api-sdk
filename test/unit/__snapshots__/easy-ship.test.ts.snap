// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`easyShip createScheduledPackage returns information about the scheduled package if structure is correct 1`] = `
Array [
  Object {
    "ScheduledPackage": Object {
      "Invoice": Object {
        "InvoiceDate": 1969-07-21T02:56:03.000Z,
        "InvoiceNumber": "String",
      },
      "PackageDimensions": Object {
        "Height": 100,
        "Length": 100,
        "Name": "String",
        "Unit": "String",
        "Width": 100,
      },
      "PackageIdentifier": "String",
      "PackagePickupSlot": Object {
        "PickupTimeEnd": 1969-07-21T02:56:03.000Z,
        "PickupTimeStart": 1969-07-21T02:56:03.000Z,
        "SlotId": "String",
      },
      "PackageStatus": "String",
      "PackageWeight": Object {
        "Unit": "String",
        "Value": 100,
      },
      "ScheduledPackageId": Object {
        "AmazonOrderId": "String",
        "PackageId": "String",
      },
    },
  },
  Object {
    "quotaMax": 1000,
    "quotaRemaining": 999,
    "quotaResetOn": 2020-04-06T10:22:23.582Z,
    "requestId": "0",
    "timestamp": 2020-05-06T09:22:23.582Z,
  },
]
`;

exports[`easyShip createScheduledPackage returns information about the scheduled package if succesful 1`] = `
Array [
  Object {
    "ScheduledPackage": Object {
      "PackageDimensions": Object {
        "Height": 12,
        "Length": 12,
        "Unit": "cm",
        "Width": 12,
      },
      "PackagePickupSlot": Object {
        "PickupTimeEnd": 2018-02-19T10:30:00.500Z,
        "PickupTimeStart": 2018-02-19T08:30:00.500Z,
        "SlotId": "Xyt123bbdg=",
      },
      "PackageStatus": "ReadyForPickup",
      "PackageWeight": Object {
        "Unit": "g",
        "Value": 23,
      },
      "ScheduledPackageId": Object {
        "AmazonOrderId": "903-1713775-3598252",
        "PackageId": "1ab0f06a-9149-87e0-aba9-7098117872d6",
      },
    },
  },
  Object {
    "quotaMax": 1000,
    "quotaRemaining": 999,
    "quotaResetOn": 2020-04-06T10:22:23.582Z,
    "requestId": "0",
    "timestamp": 2020-05-06T09:22:23.582Z,
  },
]
`;

exports[`easyShip getScheduledPackage returns info about the scheduled package if succesful 1`] = `
Array [
  Object {
    "ScheduledPackage": Object {
      "PackageDimensions": Object {
        "Height": 12,
        "Length": 12,
        "Unit": "cm",
        "Width": 12,
      },
      "PackagePickupSlot": Object {
        "PickupTimeEnd": 2018-02-19T10:30:00.500Z,
        "PickupTimeStart": 2018-02-19T08:30:00.500Z,
        "SlotId": "Xyt123bbdg=",
      },
      "PackageStatus": "Scheduled",
      "PackageWeight": Object {
        "Unit": "g",
        "Value": 23,
      },
      "ScheduledPackageId": Object {
        "AmazonOrderId": "903-1713775-3598252",
        "PackageId": "1ab0f06a-9149-87e0-aba9-7098117872d6",
      },
    },
  },
  Object {
    "quotaMax": 1000,
    "quotaRemaining": 999,
    "quotaResetOn": 2020-04-06T10:22:23.582Z,
    "requestId": "0",
    "timestamp": 2020-05-06T09:22:23.582Z,
  },
]
`;

exports[`easyShip getServiceStatus returns a parsed model when the status response is valid 1`] = `
Array [
  Object {
    "Status": "GREEN",
    "Timestamp": "2020-05-06T08:22:23.582Z",
  },
  Object {
    "quotaMax": 1000,
    "quotaRemaining": 999,
    "quotaResetOn": 2020-04-06T10:22:23.582Z,
    "requestId": "0",
    "timestamp": 2020-05-06T09:22:23.582Z,
  },
]
`;

exports[`easyShip listPickupSlots returns list of pickup slots if succesful 1`] = `
Array [
  Object {
    "AmazonOrderId": "String",
    "PickupSlotList": Array [
      Object {
        "PickupTimeEnd": 1969-07-21T02:56:03.000Z,
        "PickupTimeStart": 1969-07-21T02:56:03.000Z,
        "SlotId": "String",
      },
    ],
  },
  Object {
    "quotaMax": 1000,
    "quotaRemaining": 999,
    "quotaResetOn": 2020-04-06T10:22:23.582Z,
    "requestId": "0",
    "timestamp": 2020-05-06T09:22:23.582Z,
  },
]
`;

exports[`easyShip updateScheduledPackages returns a list of packages if succesful 1`] = `
Array [
  Object {
    "ScheduledPackageList": Array [
      Object {
        "PackageDimensions": Object {
          "Height": 12,
          "Length": 12,
          "Unit": "cm",
          "Width": 12,
        },
        "PackagePickupSlot": Object {
          "PickupTimeEnd": 2018-11-05T10:30:00.500Z,
          "PickupTimeStart": 2018-11-05T08:30:00.500Z,
          "SlotId": "Xyt123bbdg=",
        },
        "PackageStatus": "Scheduled",
        "PackageWeight": Object {
          "Unit": "g",
          "Value": 23,
        },
        "ScheduledPackageId": Object {
          "AmazonOrderId": "903-1713775-3598252",
          "PackageId": "1ab0f06a-9149-87e0-aba9-7098117872d6",
        },
      },
    ],
  },
  Object {
    "quotaMax": 1000,
    "quotaRemaining": 999,
    "quotaResetOn": 2020-04-06T10:22:23.582Z,
    "requestId": "0",
    "timestamp": 2020-05-06T09:22:23.582Z,
  },
]
`;
